### PR TITLE
add AchievementsRemaining to awardachievement API response

### PR DESCRIPTION
**Pull request by Jamiras**
Allows the server to tell the client when the user has mastered a set as part of the achievement unlock API call, instead of requiring the client make a second call to find out.

The mastery check already occurs in this API to award the mastery badge. This just modifies the logic to return the number of achievements still needed before the mastery badge would be awarded. If this value is 0, the client may choose to show a mastery notification.

This does add one additional query (to get the remaining unlocks needed) when the client asks to unlock an already unlocked achievement, but it also eliminates a query to get the game for the achievement, which was already determined earlier. So for most unlocks, there will be one less query.

The `AwardsRemaining` value will reflect the `h` parameter. if `h=1`, it will report the number of hardcore unlocks remaining. If `h=0`, it will reflect the number of non-hardcore unlocks remaining.

Before (newly awarded):
```
{"Success":true,"Score":58834,"AchievementID":40403}
```
After (newly awarded):
```
{"Success":true,"AwardsRemaining":28,"Score":58834,"AchievementID":40403}
```
Before (previously awarded):
```
{"Success":false,"Error":"User already has hardcore and regular achievements awarded.","Score":58874,"AchievementID":40403}
```
After (previously awarded):
```
{"Success":false,"AwardsRemaining":28,"Error":"User already has hardcore and regular achievements awarded.","Score":58874,"AchievementID":40403}
```
Before (newly mastered):
```
{"Success":true,"Score":58874,"AchievementID":13381}
```
After (newly mastered):
```
{"Success":true,"AwardsRemaining":0,"Score":58874,"AchievementID":13381}
```
Before (previously mastered):
```
{"Success":false,"Error":"User already has hardcore and regular achievements awarded.","Score":58874,"AchievementID":13381}
```
After (previously mastered):
```
{"Success":false,"AwardsRemaining":0,"Error":"User already has hardcore and regular achievements awarded.","Score":58874,"AchievementID":13381}
```

### CODE REVIEW

**Review by MagmarFire — DISMISSED** [review]

File: lib/database/achievement.php [hunk 1]
```diff
@@ -350,27 +351,41 @@ function addEarnedAchievementJSON($user, $achIDToAward, $isHardcore)
...
+        // the achievement unlock notification first
+        postActivity($user, ActivityType::EarnedAchivement, $achIDToAward, $isHardcore);
+    }
 
-    $awardedOK = InsertAwardedAchievementDB($user, $achIDToAward, $isHardcore);
-    if ($awardedOK && $isHardcore) {
-        $awardedOK |= InsertAwardedAchievementDB($user, $achIDToAward, false);
+    $completion = testFullyCompletedGame($achData['GameID'], $user, $isHardcore, !$alreadyAwarded);
+    if (array_key_exists('NumAwarded', $completion)) {
+        $retVal['AwardsRemaining'] = $completion['NumAch'] - $completion['NumAwarded'];
```
[the remark this answers is not in the document]
Comment (Jamiras) [reply lib/database/achievement.php:377]: That's just keeping with the internal naming, but I'm fine changing it for external use.

File: lib/database/game.php [hunk 2]
```diff
@@ -584,23 +575,23 @@ function testFullyCompletedGame($user, $achID, $isHardcore)
...
             if (!RecentlyPostedCompletionActivity($user, $gameID, $isHardcore)) {
                 postActivity($user, ActivityType::CompleteGame, $gameID, $isHardcore);
             }
-            return true;
-        } else {
-            return false;
         }
+
+        return $data;
     } else {
```
[the remark this answers is not in the document]
Comment (Jamiras) [reply lib/database/game.php:589]: Will do.

File: lib/database/achievement.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (luchaos) [reply lib/database/achievement.php]: Nice :) I like to check API responses twice as those tend to be historically hard to change after the fact.
Would you be fine with `AchievementsRemaining`? I like to think that'd be in line with v2's naming schema.

[the remark this answers is not in the document]
Comment (Jamiras) [reply lib/database/achievement.php]: Yep, already made that change: https://github.com/RetroAchievements/RAWeb/pull/827/commits/ddd358f9c6bdeca56e45b864e3666fdc2072dbed#diff-c784498cc33f282a41416b9e8e724c746487d0898c66532874d248741e951b96R377

[the remark this answers is not in the document]
Comment (luchaos) [reply lib/database/achievement.php]: Fantastic, thanks!

**Review by luchaos — APPROVED** [review]

**Pull request merged**